### PR TITLE
Add squid proxy as egress cache

### PR DESCRIPTION
**Pull request by zhouzhuojie**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62103

This PR adds a squid proxy that's deployed dedicated for PyTorch CI. Initially we only roll out to GHA, and if things are ok we will extend this to circleci tests if necessary.

`http_proxy` and `https_proxy` are compatible with the following http clients:

- curl
- wget
- python

Existing cache policy:

```
refresh_pattern -i .(7z|deb|rpm|exe|zip|tar|tgz|gz|ram|rar|bin|tiff|bz2|run|csv|sh)$ 1440 80% 2880
```

It uses the standard squid refresh_pattern for cache requests. In our setup, we tried
to cache at least (1440 minutes - 1 day) and at max (2880 minutes - 2 days), with
last-modified factor 80% ([squid doc](http://www.squid-cache.org/Doc/config/refresh_pattern/)). Please refer to [pytorch/test-infra](https://github.com/pytorch/test-infra/tree/master/aws/websites/squid-proxy) for details.

Right now, it only applies to the `build` and `test` step, to limit the scope and make sure build and test are more reliable with egress cache.

Differential Revision: [D29892919](https://our.internmc.facebook.com/intern/diff/D29892919)

### CODE REVIEW

File: .github/templates/bazel_ci_workflow.yml.j2 [hunk 1]
```diff
@@ -60,6 +60,7 @@ name: Bazel Linux CI (!{{ build_environment }})
             -e CUSTOM_TEST_ARTIFACT_BUILD_DIR \
             -e SKIP_SCCACHE_INITIALIZATION=1 \
             -e TORCH_CUDA_ARCH_LIST \
+            -e use_proxy=on http_proxy="$SQUID_PROXY" -e https_proxy="$SQUID_PROXY" -e HTTP_PROXY="$SQUID_PROXY" -e HTTPS_PROXY="$SQUID_PROXY" \
```
Comment (malfet) [inline .github/templates/bazel_ci_workflow.yml.j2:63]:
```suggestion
            -e use_proxy=on -e http_proxy="$SQUID_PROXY" -e https_proxy="$SQUID_PROXY" -e HTTP_PROXY="$SQUID_PROXY" -e HTTPS_PROXY="$SQUID_PROXY" \
```

File: .github/templates/bazel_ci_workflow.yml.j2 [hunk 2]
```diff
@@ -103,6 +104,7 @@ name: Bazel Linux CI (!{{ build_environment }})
             -e JOB_BASE_NAME \
             -e MAX_JOBS="$(nproc --ignore=2)" \
             -e SCCACHE_BUCKET \
+            -e use_proxy=on http_proxy="$SQUID_PROXY" -e https_proxy="$SQUID_PROXY" -e HTTP_PROXY="$SQUID_PROXY" -e HTTPS_PROXY="$SQUID_PROXY" \
```
Comment (malfet) [inline .github/templates/bazel_ci_workflow.yml.j2:107]:
```suggestion
            -e use_proxy=on -e http_proxy="$SQUID_PROXY" -e https_proxy="$SQUID_PROXY" -e HTTP_PROXY="$SQUID_PROXY" -e HTTPS_PROXY="$SQUID_PROXY" \
```

**Comment by malfet** [comment]
Can you please add a bit more details to PR description? I.e. why both `http_proxy` and `HTTP_PROXY` is needed. Same about `use_proxy=on`

**Review by janeyx99 — APPROVED** [review]

**Review by malfet — CHANGES_REQUESTED** [review]
I think you are missing `-e` option in bazel workflow template

**Comment by zhouzhuojie** [comment]
@zhouzhuojie has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29892919).

**Review by seemethere — APPROVED** [review]
LGTM

Note entirely sure if case matters for these calls but the Docker documentation does suggest using all uppercase for the env variable names:

Screen Shot 2021-07-26 at 11 35 10 AM

From https://docs.docker.com/network/proxy/

**Review by driazati — APPROVED** [review]
lgtm too, just curious if squid proxy is down / unreachable for some reason all network accesses will timeout / fail right?

**Comment by zhouzhuojie** [comment]
> lgtm too, just curious if squid proxy is down / unreachable for some reason all network accesses will timeout / fail right?

Yes, the proxy is in the critical path, and we will beef up the monitoring of it

**Comment by zhouzhuojie** [comment]
> LGTM
> 
> Note entirely sure if case matters for these calls but the Docker documentation does suggest using all uppercase for the env variable names:
> 
> Screen Shot 2021-07-26 at 11 35 10 AM
> 
> From https://docs.docker.com/network/proxy/

The common http clients usually respect both format, and lowercase is kind of the default. Found a good summary on this topic

https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/


image

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**